src/Debugger/Impl/ExecutionTracing/IRExecutionTracer.cs

```diff
@@ -0,0 +1,115 @@
+        Task BreakAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// When paused at a Browse> prompt, continue execution.
+        /// </summary>
+        Task ContinueAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// When paused at a Browse> prompt, step into the next call.
+        /// </summary>
+        /// <returns>
+        /// A task that completes when the step is either completed or interrupted (e.g. by a breakpoint).
+        /// The result is <see langword="true"/> if step was completed, and <see langword="false"/> if it was abandoned.
+        /// </returns>
+        /// <remarks>
+        /// Detailed semantics of step in are described in R documentation for
+        /// <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/browser.html"><c>browser()</c></a> "s" command.
+        /// </remarks>
+        Task<bool> StepIntoAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// When paused at a Browse> prompt, step over the next call.
+        /// </summary>
+        /// <returns>
+        /// A task that completes when the step is either completed or interrupted (e.g. by a breakpoint).
+        /// The result is <see langword="true"/> if step was completed, and <see langword="false"/> if it was abandoned.
+        /// </returns>
+        /// <remarks>
+        /// Detailed semantics of step over are described in R documentation for
+        /// <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/browser.html"><c>browser()</c></a> "n" command.
+        /// </remarks>
+        Task<bool> StepOverAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// When paused at a Browse> prompt, step out from the current call.
+        /// </summary>
+        /// <returns>
+        /// A task that completes when the step is either completed or interrupted (e.g. by a breakpoint).
+        /// The result is <see langword="true"/> if step was completed, and <see langword="false"/> if it was abandoned.
+        /// </returns>
+        /// Detailed semantics of step out are described in R documentation for
+        /// <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/browserText.html"><c>browserSetDebug()</c></a> function.
+        /// </remarks>
+        Task<bool> StepOutAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// If a step operation is currently in progress, cancel it.
+        /// </summary>
+        bool CancelStep();
+
+        /// <summary>
+        /// Enables or disables breakpoints. When enabled, <see cref="IRBreakpoint.BreakpointHit"/> events are raised.
+        /// </summary>
+        Task EnableBreakpointsAsync(bool enable, CancellationToken cancellationToken = default(CancellationToken));
```

Comment:
Toggle implies changing state to the reverse of what it's currently is (and so requires you to be aware of what the current state is). This one actually lets you do EnableBreakpointsAsync(true) regardless of what the current state is - and if they're already enabled, it'll do nothing.

---
In reply to: [62136887](https://github.com/Microsoft/RTVS/pull/1731#discussion_r62136887) [](ancestors = 62136887)